diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index 386ec75..ac2bba2 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -418,7 +418,10 @@ def _takes_container(factory: Callable) -> bool:
     if name == "svcs_container":
         return True
 
-    if (annot := p.annotation) is Container or annot == "svcs.Container":
+    if (
+        (annot := p.annotation) is Container
+        or annot in {"svcs.Container", Container.__name__}
+    ):
         return True
 
     return False
diff --git a/tests/test_registry.py b/tests/test_registry.py
index 4b9370f..2231dd3 100644
--- a/tests/test_registry.py
+++ b/tests/test_registry.py
@@ -358,6 +358,20 @@ class TestTakesContainer:
 
         assert svcs._core._takes_container(factory)
 
+    def test_annotation_str_short(self):
+        """
+        Return true if the first argument is annotated as `Container` using a
+        string.
+
+        This happens under ``from __future__ import annotations`` when the
+        user writes ``from svcs import Container``.
+        """
+
+        def factory(bar: "Container"):
+            ...
+
+        assert svcs._core._takes_container(factory)
+
     def test_catches_invalid_sigs(self):
         """
         If the factory takes more than one parameter, raise an TypeError.
