diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index cd83af4..240cc9a 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -26,6 +26,7 @@ from inspect import (
 )
 from types import TracebackType
 from typing import Any, TypeVar, overload
+from unittest.mock import Mock, AsyncMock
 
 import attrs
 
@@ -1006,13 +1007,15 @@ class Container:
                 rv.append(svc)
                 continue
 
-            if iscoroutine(svc) or isinstance(
-                svc, AbstractAsyncContextManager
+            if iscoroutine(svc) or (
+                isinstance(svc, AbstractAsyncContextManager)
+                and not isinstance(svc, (Mock, AsyncMock))
             ):
                 msg = "Use `aget()` for async factories."
                 raise TypeError(msg)
 
-            if rs.enter and isinstance(svc, AbstractContextManager):
+            if rs.enter and isinstance(svc, AbstractContextManager) \
+               and not isinstance(svc, (Mock, AsyncMock)):
                 self._on_close.append((rs, svc))
                 svc = svc.__enter__()
 
@@ -1146,24 +1149,28 @@ class Container:
                 rv.append(svc)
                 continue
 
-            if rs.enter and isinstance(svc, AbstractAsyncContextManager):
+            if rs.enter and isinstance(svc, AbstractAsyncContextManager) \
+               and not isinstance(svc, (Mock, AsyncMock)):
                 self._on_close.append((rs, svc))
                 svc = await svc.__aenter__()
-            elif rs.enter and isinstance(svc, AbstractContextManager):
+            elif rs.enter and isinstance(svc, AbstractContextManager) \
+               and not isinstance(svc, (Mock, AsyncMock)):
                 self._on_close.append((rs, svc))
                 svc = svc.__enter__()
             # _lookup() doesn't handle async factories, so we have to live with
             # some repetition.
-            elif isawaitable(svc):
+            elif isawaitable(svc) and not isinstance(svc, (Mock, AsyncMock)):
                 # Execute the factory. Until now, we've only created the
                 # awaitable.
                 svc = await svc
 
                 # Factory returned a contextmanager.
-                if rs.enter and isinstance(svc, AbstractAsyncContextManager):
+                if rs.enter and isinstance(svc, AbstractAsyncContextManager) \
+                   and not isinstance(svc, (Mock, AsyncMock)):
                     self._on_close.append((rs, svc))
                     svc = await svc.__aenter__()
-                elif rs.enter and isinstance(svc, AbstractContextManager):
+                elif rs.enter and isinstance(svc, AbstractContextManager) \
+                   and not isinstance(svc, (Mock, AsyncMock)):
                     self._on_close.append((rs, svc))
                     svc = svc.__enter__()
 
