diff --git a/src/svcs/fastapi.py b/src/svcs/fastapi.py
index 3ab361d..770873e 100644
--- a/src/svcs/fastapi.py
+++ b/src/svcs/fastapi.py
@@ -8,7 +8,7 @@ import contextlib
 import inspect
 
 from collections.abc import AsyncGenerator, Callable
-from typing import Annotated, TypeAlias, cast
+from typing import Annotated, TypeAlias, cast, Any
 
 import attrs
 
diff --git a/src/svcs/starlette.py b/src/svcs/starlette.py
index db1dd24..6858fb1 100644
--- a/src/svcs/starlette.py
+++ b/src/svcs/starlette.py
@@ -262,3 +262,20 @@ async def aget(request: Request, *svc_types: _ServiceType) -> object:
     Same as :meth:`svcs.Container.aget`, but uses the container from *request*.
     """
     return await svcs_from(request).aget(*svc_types)
+
+
+def get_registry(app_or_client: Any) -> svcs.Registry:
+    app = getattr(app_or_client, "app", app_or_client)
+    state = getattr(app, "state", None)
+    if state is None:
+        raise RuntimeError("The provided object does not have a 'state' attribute.")
+
+    registry = getattr(state, _KEY_REGISTRY, None)
+    if registry is None and isinstance(state, dict):
+        registry = state.get(_KEY_REGISTRY)
+
+    if registry is None:
+        raise RuntimeError("No registry attached to the application.")
+    if not registry._services:
+        raise RuntimeError("The registry has been closed.")
+    return registry
