diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index cd83af4..756d845 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -9,6 +9,8 @@ import logging
 import sys
 import warnings
 
+import unittest.mock
+
 from collections.abc import Awaitable, Callable, Iterator
 from contextlib import (
     AbstractAsyncContextManager,
@@ -32,6 +34,10 @@ import attrs
 from .exceptions import ServiceNotFoundError
 
 
+def _is_mock(obj: Any) -> bool:
+    return isinstance(obj, unittest.mock.Mock)
+
+
 log = logging.getLogger("svcs")
 
 
@@ -1006,15 +1012,16 @@ class Container:
                 rv.append(svc)
                 continue
 
-            if iscoroutine(svc) or isinstance(
-                svc, AbstractAsyncContextManager
-            ):
-                msg = "Use `aget()` for async factories."
-                raise TypeError(msg)
+            if not _is_mock(svc):
+                if iscoroutine(svc) or isinstance(
+                    svc, AbstractAsyncContextManager
+                ):
+                    msg = "Use `aget()` for async factories."
+                    raise TypeError(msg)
 
-            if rs.enter and isinstance(svc, AbstractContextManager):
-                self._on_close.append((rs, svc))
-                svc = svc.__enter__()
+                if rs.enter and isinstance(svc, AbstractContextManager):
+                    self._on_close.append((rs, svc))
+                    svc = svc.__enter__()
 
             self._instantiated[svc_type] = (svc, rs)
 
@@ -1146,26 +1153,27 @@ class Container:
                 rv.append(svc)
                 continue
 
-            if rs.enter and isinstance(svc, AbstractAsyncContextManager):
-                self._on_close.append((rs, svc))
-                svc = await svc.__aenter__()
-            elif rs.enter and isinstance(svc, AbstractContextManager):
-                self._on_close.append((rs, svc))
-                svc = svc.__enter__()
-            # _lookup() doesn't handle async factories, so we have to live with
-            # some repetition.
-            elif isawaitable(svc):
-                # Execute the factory. Until now, we've only created the
-                # awaitable.
-                svc = await svc
-
-                # Factory returned a contextmanager.
+            if not _is_mock(svc):
                 if rs.enter and isinstance(svc, AbstractAsyncContextManager):
                     self._on_close.append((rs, svc))
                     svc = await svc.__aenter__()
                 elif rs.enter and isinstance(svc, AbstractContextManager):
                     self._on_close.append((rs, svc))
                     svc = svc.__enter__()
+                # _lookup() doesn't handle async factories, so we have to live with
+                # some repetition.
+                elif isawaitable(svc):
+                    # Execute the factory. Until now, we've only created the
+                    # awaitable.
+                    svc = await svc
+
+                    # Factory returned a contextmanager.
+                    if rs.enter and isinstance(svc, AbstractAsyncContextManager):
+                        self._on_close.append((rs, svc))
+                        svc = await svc.__aenter__()
+                    elif rs.enter and isinstance(svc, AbstractContextManager):
+                        self._on_close.append((rs, svc))
+                        svc = svc.__enter__()
 
             self._instantiated[svc_type] = (svc, rs)
 
