diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index 968b0f4..4864140 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -720,6 +720,19 @@ class Container:
         Returns:
             A list of services that have registered a ping callable.
         """
+        pings: dict[type, RegisteredService] = {
+            rs.svc_type: rs
+            for rs in self.registry._services.values()
+            if rs.ping is not None
+        }
+
+        if self._lazy_local_registry is not None:
+            for rs in self._lazy_local_registry._services.values():
+                if rs.ping is not None:
+                    pings[rs.svc_type] = rs
+                else:
+                    pings.pop(rs.svc_type, None)
+
         return [
             ServicePing(
                 rs.name,
@@ -728,8 +741,7 @@ class Container:
                 rs.ping,
                 self,
             )
-            for rs in self.registry._services.values()
-            if rs.ping is not None
+            for rs in pings.values()
         ]
 
     def get_abstract(self, *svc_types: type) -> Any:
