diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index cd83af4..c4d9bb4 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -530,6 +530,14 @@ else:
             return None
 
 
+def _is_mock(obj: Any) -> bool:
+    try:
+        from unittest.mock import MagicMock, AsyncMock
+        return isinstance(obj, (MagicMock, AsyncMock))
+    except ImportError:
+        return False
+
+
 def _takes_container(factory: Callable) -> bool:
     """
     Return True if *factory* takes a svcs.Container as its first argument.
@@ -1006,6 +1014,11 @@ class Container:
                 rv.append(svc)
                 continue
 
+            if _is_mock(svc):
+                self._instantiated[svc_type] = (svc, rs)
+                rv.append(svc)
+                continue
+
             if iscoroutine(svc) or isinstance(
                 svc, AbstractAsyncContextManager
             ):
@@ -1146,6 +1159,11 @@ class Container:
                 rv.append(svc)
                 continue
 
+            if _is_mock(svc):
+                self._instantiated[svc_type] = (svc, rs)
+                rv.append(svc)
+                continue
+
             if rs.enter and isinstance(svc, AbstractAsyncContextManager):
                 self._on_close.append((rs, svc))
                 svc = await svc.__aenter__()
