diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index cd83af4..6d04d8c 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -42,6 +42,16 @@ def _full_name(obj: object) -> str:
         return repr(obj)
 
 
+try:
+    import unittest.mock
+    _MOCK_TYPES = (unittest.mock.Mock, unittest.mock.AsyncMock)
+except ImportError:
+    _MOCK_TYPES = ()
+
+def _is_mock(obj: Any) -> bool:
+    return isinstance(obj, _MOCK_TYPES)
+
+
 # Default names where to put the container and registry in integrations.
 _KEY_REGISTRY = "svcs_registry"
 _KEY_CONTAINER = "svcs_container"
@@ -1006,13 +1016,13 @@ class Container:
                 rv.append(svc)
                 continue
 
-            if iscoroutine(svc) or isinstance(
+            if not _is_mock(svc) and (iscoroutine(svc) or isinstance(
                 svc, AbstractAsyncContextManager
-            ):
+            )):
                 msg = "Use `aget()` for async factories."
                 raise TypeError(msg)
 
-            if rs.enter and isinstance(svc, AbstractContextManager):
+            if rs.enter and isinstance(svc, AbstractContextManager) and not _is_mock(svc):
                 self._on_close.append((rs, svc))
                 svc = svc.__enter__()
 
@@ -1146,24 +1156,24 @@ class Container:
                 rv.append(svc)
                 continue
 
-            if rs.enter and isinstance(svc, AbstractAsyncContextManager):
+            if rs.enter and isinstance(svc, AbstractAsyncContextManager) and not _is_mock(svc):
                 self._on_close.append((rs, svc))
                 svc = await svc.__aenter__()
-            elif rs.enter and isinstance(svc, AbstractContextManager):
+            elif rs.enter and isinstance(svc, AbstractContextManager) and not _is_mock(svc):
                 self._on_close.append((rs, svc))
                 svc = svc.__enter__()
             # _lookup() doesn't handle async factories, so we have to live with
             # some repetition.
-            elif isawaitable(svc):
+            elif isawaitable(svc) and not _is_mock(svc):
                 # Execute the factory. Until now, we've only created the
                 # awaitable.
                 svc = await svc
 
                 # Factory returned a contextmanager.
-                if rs.enter and isinstance(svc, AbstractAsyncContextManager):
+                if rs.enter and isinstance(svc, AbstractAsyncContextManager) and not _is_mock(svc):
                     self._on_close.append((rs, svc))
                     svc = await svc.__aenter__()
-                elif rs.enter and isinstance(svc, AbstractContextManager):
+                elif rs.enter and isinstance(svc, AbstractContextManager) and not _is_mock(svc):
                     self._on_close.append((rs, svc))
                     svc = svc.__enter__()
 
