diff --git a/docs/core-concepts.md b/docs/core-concepts.md
index d0e6960..dbfbfb4 100644
--- a/docs/core-concepts.md
+++ b/docs/core-concepts.md
@@ -313,7 +313,7 @@ Now, you can point your monitoring tool of choice -- like Prometheus's [Blackbox
 
 While *svcs*'s core is entirely agnostic on how you use the registry and the container, all our {doc}`integrations/index` follow the same life cycle:
 
-- The {class}`svcs.Registry` objects live on **application-scoped** objects like {attr}`flask.Flask.config`.
+- The {class}`svcs.Registry` objects live on **application-scoped** objects like {attr}`flask.Flask.extensions`.
 - The {class}`svcs.Container` objects live on **request-scoped** objects like {data}`flask.g`.
 
 You're free to structure your own integrations as you want, though.
diff --git a/docs/integrations/flask.md b/docs/integrations/flask.md
index 9497a7b..513b84b 100644
--- a/docs/integrations/flask.md
+++ b/docs/integrations/flask.md
@@ -1,6 +1,6 @@
 # Flask
 
-*svcs*'s [Flask](https://flask.palletsprojects.com/) integration uses the {attr}`flask.Flask.config` object to store the {class}`svcs.Registry` and the {obj}`~flask.g` object to store the {class}`svcs.Container`.
+*svcs*'s [Flask](https://flask.palletsprojects.com/) integration uses the {attr}`flask.Flask.extensions` object to store the {class}`svcs.Registry` and the {obj}`~flask.g` object to store the {class}`svcs.Container`.
 It also installs a {meth}`flask.Flask.teardown_appcontext` handler to close the container when the request is done.
 
 ---
diff --git a/docs/why.md b/docs/why.md
index 8b233ea..7735fa3 100644
--- a/docs/why.md
+++ b/docs/why.md
@@ -166,7 +166,7 @@ If you're curious, check the [glossary](glossary) entries for {term}`Service Loc
 While it may take a moment to realize, all of this comes with many benefits:
 
 **Reduction in boilerplate.**
-Every web framework has some way to store data on long-lived objects like the application, and short-lived objects like requests (for example, in Flask it's {attr}`flask.Flask.config` and {obj}`flask.g` – Starlette uses `request.state` for both).
+Every web framework has some way to store data on long-lived objects like the application, and short-lived objects like requests (for example, in Flask it's {attr}`flask.Flask.extensions` and {obj}`flask.g` – Starlette uses `request.state` for both).
 Some frameworks also have helpers to control the lifecycle of those objects (like {meth}`flask.Flask.teardown_appcontext` or {meth}`pyramid.request.Request.add_finished_callback`).
 But they work subtly differently and you accumulate a lot of repetitive boilerplate code.
 In fact, Hynek started this project because of the [repetitiveness of Flask's `get_X` pattern](#flask-get-x).
diff --git a/src/svcs/flask.py b/src/svcs/flask.py
index 810f74c..f999530 100644
--- a/src/svcs/flask.py
+++ b/src/svcs/flask.py
@@ -35,7 +35,7 @@ def svcs_from(g: _AppCtxGlobals = g) -> Container:
     Get the current container from *g*.
     """
     if (con := g.get(_KEY_CONTAINER, None)) is None:
-        con = Container(current_app.config[_KEY_REGISTRY])
+        con = Container(current_app.extensions[_KEY_REGISTRY])
         setattr(g, _KEY_CONTAINER, con)
 
     return con  # type: ignore[no-any-return]
@@ -53,7 +53,7 @@ def get_registry(app: Flask | None = None) -> Registry:
     """
     if app is None:
         app = current_app
-    return app.config[_KEY_REGISTRY]  # type: ignore[no-any-return]
+    return app.extensions[_KEY_REGISTRY]  # type: ignore[no-any-return]
 
 
 registry = cast(Registry, LocalProxy(get_registry))
@@ -68,7 +68,7 @@ def init_app(app: FlaskAppT, *, registry: Registry | None = None) -> FlaskAppT:
 
     Creates a registry for you if you don't provide one.
     """
-    app.config[_KEY_REGISTRY] = registry or Registry()
+    app.extensions[_KEY_REGISTRY] = registry or Registry()
     app.teardown_appcontext(teardown)
 
     return app
@@ -95,7 +95,7 @@ def register_factory(
     Same as :meth:`svcs.Registry.register_factory()`, but uses registry on
     *app* that has been put there by :func:`init_app()`.
     """
-    app.config[_KEY_REGISTRY].register_factory(
+    app.extensions[_KEY_REGISTRY].register_factory(
         svc_type,
         factory,
         enter=enter,
@@ -117,7 +117,7 @@ def register_value(
     Same as :meth:`svcs.Registry.register_value()`, but uses registry on *app*
     that has been put there by :func:`init_app()`.
     """
-    app.config[_KEY_REGISTRY].register_value(
+    app.extensions[_KEY_REGISTRY].register_value(
         svc_type,
         value,
         enter=enter,
@@ -209,7 +209,7 @@ def close_registry(app: Flask) -> None:
     """
     Close the registry on *app*, if present.
     """
-    if reg := app.config.pop(_KEY_REGISTRY, None):
+    if reg := app.extensions.pop(_KEY_REGISTRY, None):
         reg.close()
 
 
diff --git a/tests/integrations/test_flask.py b/tests/integrations/test_flask.py
index abcca4c..43a3153 100644
--- a/tests/integrations/test_flask.py
+++ b/tests/integrations/test_flask.py
@@ -263,7 +263,7 @@ class TestInitApp:
         app = flask.Flask("tests")
         svcs.flask.init_app(app)
 
-        assert isinstance(app.config["svcs_registry"], svcs.Registry)
+        assert isinstance(app.extensions["svcs_registry"], svcs.Registry)
 
     def test_explicit_registry(self):
         """
@@ -273,7 +273,7 @@ class TestInitApp:
         app = flask.Flask("tests")
         svcs.flask.init_app(app, registry=registry)
 
-        assert registry is app.config["svcs_registry"]
+        assert registry is app.extensions["svcs_registry"]
 
 
 class TestCloseRegistry:
