diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index 4ed8a51..d90c595 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -1011,9 +1011,21 @@ class Container:
             elif enter and isinstance(svc, AbstractContextManager):
                 self._on_close.append((name, svc))
                 svc = svc.__enter__()
+            # _lookup() doesn't handle async factories, so we have to live with
+            # some repetition.
             elif isawaitable(svc):
+                # Execute the factory. Until now, we've only created the
+                # awaitable.
                 svc = await svc
 
+                # Factory returned a contextmanager.
+                if enter and isinstance(svc, AbstractAsyncContextManager):
+                    self._on_close.append((name, svc))
+                    svc = await svc.__aenter__()
+                elif enter and isinstance(svc, AbstractContextManager):
+                    self._on_close.append((name, svc))
+                    svc = svc.__enter__()
+
             self._instantiated[svc_type] = svc
 
             rv.append(svc)
diff --git a/tests/test_container.py b/tests/test_container.py
index 4f9838b..12dda6c 100644
--- a/tests/test_container.py
+++ b/tests/test_container.py
@@ -4,6 +4,7 @@
 
 import gc
 
+from contextlib import asynccontextmanager, contextmanager
 from unittest.mock import Mock
 
 import pytest
@@ -91,6 +92,56 @@ class TestContainer:
 
         assert close_me.is_aclosed
 
+    @pytest.mark.asyncio
+    async def test_async_context_manager_sync_factory(self, container):
+        """
+        If an async factory returns an context manager, it's entered.
+        """
+        entered = left = False
+
+        @contextmanager
+        def cm():
+            nonlocal entered, left
+            entered = True
+            yield 42
+            left = True
+
+        async def factory(svcs_container):
+            return cm()
+
+        container.registry.register_factory(int, factory)
+
+        async with container:
+            assert 42 == await container.aget(int)
+
+        assert entered
+        assert left
+
+    @pytest.mark.asyncio
+    async def test_async_context_manager_async_factory(self, container):
+        """
+        If an async factory returns an async context manager, it's entered.
+        """
+        entered = left = False
+
+        @asynccontextmanager
+        async def cm():
+            nonlocal entered, left
+            entered = True
+            yield 42
+            left = True
+
+        async def factory(svcs_container):
+            return cm()
+
+        container.registry.register_factory(int, factory)
+
+        async with container:
+            assert 42 == await container.aget(int)
+
+        assert entered
+        assert left
+
     def test_gc_warning(self, recwarn, registry):
         """
         If a container is gc'ed with pending cleanups, a warning is raised.
