diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index cd83af4..955422c 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -26,6 +26,7 @@ from inspect import (
 )
 from types import TracebackType
 from typing import Any, TypeVar, overload
+from unittest.mock import MagicMock
 
 import attrs
 
@@ -1006,8 +1007,9 @@ class Container:
                 rv.append(svc)
                 continue
 
-            if iscoroutine(svc) or isinstance(
-                svc, AbstractAsyncContextManager
+            if not isinstance(svc, MagicMock) and (
+                iscoroutine(svc)
+                or isinstance(svc, AbstractAsyncContextManager)
             ):
                 msg = "Use `aget()` for async factories."
                 raise TypeError(msg)
diff --git a/tests/test_integration.py b/tests/test_integration.py
index cbe4a15..a9c5235 100644
--- a/tests/test_integration.py
+++ b/tests/test_integration.py
@@ -12,6 +12,7 @@ from contextlib import (
     contextmanager,
 )
 from typing import NewType
+from unittest.mock import MagicMock
 
 import pytest
 
@@ -258,6 +259,40 @@ def test_get_on_async_factory_raises_type_error(
         ) == recwarn.pop().message.args
 
 
+def test_get_with_magicmock_factory_works(registry, container):
+    """
+    A sync factory can return a MagicMock as long as they're registered with
+    enter=False.
+    """
+
+    mm = MagicMock()
+    registry.register_factory(Service, lambda: mm, enter=False)
+
+    svc = container.get(Service)
+
+    assert mm is svc
+    assert svc is container.get(Service)
+
+
+@pytest.mark.asyncio
+async def test_aget_with_asyncmock_factory_works(registry, container):
+    """
+    A async factory can return a MagicMock as long as they're registered with
+    enter=False.
+    """
+    mm = MagicMock()
+
+    async def factory():
+        return mm
+
+    registry.register_factory(Service, factory, enter=False)
+
+    svc = await container.aget(Service)
+
+    assert mm is svc
+    assert svc is container.get(Service)
+
+
 def test_local_value_overrides_global_value(registry, container):
     """
     If a container registers a local value, it takes precedence of the global
