diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index d90c595..2354d83 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -8,7 +8,7 @@ import inspect
 import logging
 import warnings
 
-from collections.abc import Awaitable, Callable
+from collections.abc import Awaitable, Callable, Iterator
 from contextlib import (
     AbstractAsyncContextManager,
     AbstractContextManager,
@@ -48,6 +48,26 @@ _KEY_CONTAINER = "svcs_container"
 
 @attrs.frozen
 class RegisteredService:
+    """
+    A recipe for creating a service.
+
+    .. warning::
+
+        Strictly read-only.
+
+    Attributes:
+        svc_type: The type under which the type has been registered.
+
+        factory: Callable that creates the service.
+
+        takes_container:
+            Whether the factory takes a container as its first argument.
+
+        enter: Whether context managers returned by the factory are entered.
+
+        ping: See :ref:`health`.
+    """
+
     svc_type: type
     factory: Callable = attrs.field(hash=False)
     takes_container: bool
@@ -161,6 +181,15 @@ class Registry:
         """
         return svc_type in self._services
 
+    def __iter__(self) -> Iterator[RegisteredService]:
+        """
+        Returns:
+            An iterator over registered services.
+
+        .. versionadded:: 24.2.0
+        """
+        return iter(self._services.values())
+
     def __enter__(self) -> Registry:
         return self
 
diff --git a/tests/test_registry.py b/tests/test_registry.py
index bc6d01c..a11c266 100644
--- a/tests/test_registry.py
+++ b/tests/test_registry.py
@@ -281,6 +281,20 @@ class TestRegistry:
         assert Service in registry
         assert AnotherService not in registry
 
+    def test_iterate(self, registry):
+        """
+        It's possible to iterate over the registered services.
+        """
+        registry.register_factory(Service, Service)
+        registry.register_factory(AnotherService, AnotherService)
+
+        assert {
+            svcs.RegisteredService(Service, Service, False, True, None),
+            svcs.RegisteredService(
+                AnotherService, AnotherService, False, True, None
+            ),
+        } == {rs for rs in registry}  # noqa: C416 -- explicit on purpose
+
     def test_gc_warning(self, recwarn):
         """
         If a registry is gc'ed with pending cleanups, a warning is raised.
