diff --git a/src/svcs/_core.py b/src/svcs/_core.py
index d90c595..e3f3c0d 100644
--- a/src/svcs/_core.py
+++ b/src/svcs/_core.py
@@ -146,20 +146,18 @@ class Registry:
     def __repr__(self) -> str:
         return f"<svcs.Registry(num_services={len(self._services)})>"
 
-    def __contains__(self, svc_type: type) -> bool:
+    def __iter__(self) -> iter[RegisteredService]:
         """
-        Check whether this registry knows how to create *svc_type*:
+        Yield the registered services.
 
         .. doctest::
 
             >>> reg = svcs.Registry()
             >>> reg.register_value(int, 42)
-            >>> int in reg
-            True
-            >>> str in reg
-            False
+            >>> list(reg)
+            [<RegisteredService(svc_type=int, factory=<lambda>, takes_container=False, enter=False, has_ping=False)>]
         """
-        return svc_type in self._services
+        yield from self._services.values()
 
     def __enter__(self) -> Registry:
         return self
